### PR TITLE
Clone graph keep their original prevs nodes and next nodes sequence

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

Clone graph keep their original prevs nodes and next nodes order.
Current cloneGraph may change the next nodes order, thus will influence the update of tree-lstm (it has many shared component)

In the meantime, fix a bug in graph backward that caused by the following ut

"Graph backward" should "be correct when contains nested output"

## How was this patch tested?

unit test
test over tree lstm training

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/1743

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DirectedGraph.scala [hunk 1]
```diff
@@ -148,12 +148,17 @@ class DirectedGraph[T](val source : Node[T], val reverse : Boolean = false) exte
     })
     bfs.foreach(node => {
       if (reverseEdge) {
-        node.prevNodesAndEdges.foreach(prevNodeAndEdge => {
-          oldToNew.get(node).add(oldToNew.get(prevNodeAndEdge._1), prevNodeAndEdge._2)
+        node.nextNodesAndEdges.foreach(nextNodeAndEdge => {
+          // some node may be ignored
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DirectedGraph.scala]: Please add more comments here. e.g.
// keep the order in the nextNodes array in the current node. As we go through all node in bfs, the prevNodes order should be preserved.
// Some next nodes may be not included in the graph

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 2]
```diff
@@ -509,14 +509,17 @@ class Graph[T: ClassTag](val inputs : Seq[ModuleNode[T]],
         activity.toTensor[T].add(other.toTensor[T])
       } else {
         val actTable = activity.toTable
-        val actLen = actTable.length()
         val otherTable = other.toTable
-        val otherLen = otherTable.length()
-        require(actLen == otherLen, "table length is not equal")
+        val maxActIndex = actTable.keySet.map(_.asInstanceOf[Int]).max
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: try to use length instead of keySet.map

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by yiheng** [comment]
LGTM. Please merge it and backport to branch-0.3

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 2 above)
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: otherTable.keySet.foreach?

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by yiheng** [comment]
LGTM. Please merge and backport

**Pull request merged**